binary-data-formats.qmd

```diff
@@ -0,0 +1,404 @@
+
+## Other topics
+
+I haven't quite worked out the best way to integrate these sections.
+
+### Using duckdb with arrow
+
+One useful feature of arrow is that it is interoperable with duckdb and dbplyr.
+It is possible to open a multi-file data set using arrow and then pass it to duckdb using the `to_duckdb()` function.
+Doing so means that your dplyr code will now be interpreted by dbplyr (rather than arrow) and executed by duckdb.
+
+```{r use-duckdb}
+open_dataset("data/seattle-library-checkouts/") |>
+  to_duckdb() |>
+  filter(CheckoutYear == 2021, MaterialType == "BOOK") |>
+  group_by(CheckoutMonth) |>
+  summarise(TotalCheckouts = sum(Checkouts)) |>
+  arrange(desc(CheckoutMonth)) |>
+  collect()
+```
+
+There is an analogous function `to_arrow()` that passes control back to arrow, meaning that it is possible to swap back and forth between the duckdb and arrow back ends within a single analysis pipeline.
+This operation is extremely efficient (only the metadata need to be copied from one system to the other) can occasionally be valuable when one system implements functionality that the other does not.
+
+### The arrow in-memory and on-disk format
+
+The main focus in this chapter has been on the parquet format, but it is worth noting in passing that arrow defines its own binary data format that represents data on-disk in (almost) the exact format that it uses when representing rectangular data in-memory.
+To unpack this a little, let's load the parquet file storing the 2021 Seattle library data:
+
+```{r read-arrow-table}
+"!file.exists('data/seattle-library-checkouts/CheckoutYear=2021/part-0.parquet')"
+checkouts2021 <- read_parquet(
+  file = "data/seattle-library-checkouts/CheckoutYear=2021/part-0.parquet",
+  as_data_frame = FALSE
+)
+glimpse(checkouts2021)
+```
+
+Note that the first line of output tells us this is a Table object.
+A Table is the Apache Arrow analog to a data frame: it is represented in-memory, and arrow computations are done using these objects.
+For our purposes, the important thing to understand is that the arrow file format is designed to be an exact on-disk analog of an in-memory Table.
+You can read and write these files using the `read_ipc_file()` and `write_ipc_file()` functions in the arrow package:
+
+```{r write-arrow-file}
+arrow_file <- tempfile(fileext = ".arrow")
+write_ipc_file(checkouts2021, arrow_file)
+```
+
+Note that for future releases of arrow, these functions are likely to deprecated in favour of the names `read_arrow()` and `write_arrow()` respectively.
+Also note that the arrow file format was previously known as the "feather" file format, and you are likely to see references to it by that name.
+
+To learn more about Datasets, Tables, and other arrow data objects, see the documentation on the arrow package website: [arrow.apache.org/docs/r/](https://arrow.apache.org/docs/r/).
+
+### Data in the cloud
```

Comment:
If we keep this section (personally, I would vote yes, but not sure what the tradeoffs are at this point...) it would be nice to also talk about how the connection of datasets being able to query only the parts of parquet files that it needs to + S3 makes for a shockingly efficient setup for processing data out of S3.

Reply:
Could you expand on this a bit? Are you saying that arrow will read just the necessary bytes from the parquet header so it can only read the columns you request? Presumably it can only return a subset of rows the rows if the parquet file is sharded by the variable you're filtering on?

But doesn't this all have to happen a second time if you re-run the code? Or does arrow somehow magically cache the data locally?

Reply:
Yeah to most of what you said:

> Are you saying that arrow will read just the necessary bytes from the parquet header so it can only read the columns you request?

Yeah, one of the parts of registering a parquet dataset in S3 is reading the headers of (some of) the files (it will read all of them if you tell it to unify schemas, otherwise it only reads the first — or perhaps the first of each shard). Then when you go to actually scan the files, it does again read the header(s) and grabs only the columns that you specify.

> Presumably it can only return a subset of rows the rows if the parquet file is sharded by the variable you're filtering on?

If the Parquet file has its rows split up into row groups (which Arrow will do by default if the files are large enough), it can use the row group statistics to discard whole chunks of row groups at a time if they don't meet a specific filter. Like you allude to, that means you'll only get a benefit if there are row groups that one can ignore — which might take careful writing of the data in the first place like with sharding. Though I've found it very useful in cases where I'm filtering out rows that contain `NA`s|`NULL`s which happen to be distributed unevenly throughout the dataset (but frequently enough that many whole row groups will be all `NA` for that column and thus can be ignored.

As for cache: I don't know of Arrow doing anything special here — so yeah if you query again it'll do the whole thing over. This workflow is good for the kinds of one-shot queries and then having something local in memory that you work with after that, but not for repeated querying of the dataset like that.

Reply:
Thanks for the explanation!

Reply:
I guess what I'm wondering with this is how much is worth saying in the R4DS chapter and how much do we want to push into the Arrow docs? To me it feels like most of @jonkeane's explanation is something that should be added into the "working with cloud storage" vignette on the Arrow site. I don't know if we want to be going that deep in R4DS? (Probably @hadley's call!) 

I guess the prpblem at the moment is that on the Arrow docs it's not at all clear to users that this advantage exists, and (unfortunately) this isn't one of the vignettes I worked much on in https://github.com/apache/arrow/pull/14514/. I could open a ticket to expand on the vignette there perhaps?